README.md
# responsify

A new Flutter package for creating Responsive UI based on device type

## Overview of this library

* The ```responsify``` package is a pure Dart package based on [this](https://medium.com/flutter-community/the-best-flutter-responsive-ui-pattern-ba52875d70cd) tutorial by FilledStacks.
* With this library, you can:<BR>
  * Get the device type (Mobile, Tablet, Computer or Wearable) and create different layouts depending on device type<BR>
  * Get the device's orientation<BR>
  * Get the device's platform brightness, i.e., light mode or dark mode<BR>
  * Get the parent widget's size (Width X Height)<BR>
  * Get the child widget's size (Width X Height)<BR>
  * Get the device operating system as a String<BR>
  
## How this library works

* The library has a single widget, known as the ```ResponsiveUiWidget``` that returns a widget in its ```build``` method<BR>
* The library takes the device's pixel density into account. It is determined using the device's pixel ratio and a baseline pixel density
  of 160. Older computers have a lower pixel density (such as 96 or 120 ppi). In this case, the pixel density is not
  considered when determining if the device is a computer<BR>
* The ```ResponsiveUiWidget``` has a required parameter called ```targetOlderComputers```. If true, this flag allows you to 
  target older computer models that have pixel density of lower than 160. Also, when true, only this flag and the computer's 
  Platform OS (Windows, macOS, Linux) is used to determine if the device is a computer. If false, the device type is determined to be a computer
  using the Pythogoras theorem (see the next point) along with the Platform OS<BR>
* The Pythogoras theorem is used to calculated the diagonal dimensions of the device screen in pixels. This resulting value is then
  used with the device pixel density to determine the diagonal screen size in inches. Device Screen Size (inches) = Device Screen Size (pixels)/Device pixel density (PPI).
  This result, along with the device's Platform OS, is used to determine the device type (Mobile, Tablet, Wearable, Computer)
* If the above three points sound confusing, here is the code block:<BR>
```
    ///The checkHandheldDevicePlatform flag is used to determine a Handheld device's OS Platform
    bool checkHandheldDevicePlatform = Platform.isAndroid || Platform.isIOS || Platform.isFuchsia;

    ///The checkComputerPlatform flag is used to determine a Computer device's OS Platform
    ///Here a Computer may be any Desktop or Laptop/Notebook device that runs either Windows,
    ///macOS or Linux
    bool checkComputerPlatform = Platform.isWindows || Platform.isMacOS || Platform.isLinux;

    ///if...else block used to determine the device type based on the DeviceType enum and the
    ///DeviceInformation class. All threshold values are in inches
    if (deviceSize! <= 3) {
      return deviceType = DeviceTypeInformation.WEARABLE;
    } else if (deviceSize! > 3 && deviceSize! < 7.1 && checkHandheldDevicePlatform) {
      return deviceType = DeviceTypeInformation.MOBILE;
    } else if (deviceSize! > 7.1 && deviceSize! < 13.5 && checkHandheldDevicePlatform) {
      return deviceType = DeviceTypeInformation.TABLET;
    } else if ((deviceSize! > 13.5 && deviceSize! < 55 && checkComputerPlatform) ||
        (targetOlderComputers! && checkComputerPlatform)) {
      return deviceType = DeviceTypeInformation.COMPUTER;
    } else {
      return deviceType = DeviceTypeInformation.UNKNOWN;
    }
```
* Device pixel density = Device pixel ratio * 160. The Device Pixel Ratio is retrieved using MediaQuery<BR>
* The thresholds for device screen sizes were determined by performing a search for minimum and maximum screen sizes for each device type. 
  As such, you do not have to specify thresholds for device sizes<BR>

## Usage

1) Add ```responsify``` to your ```pubspec.yaml``` and run ```pub get```<BR>
2) Import ```responsify.dart``` into your project, like so:<BR>
   ```import 'package:responsify/responsify.dart';'```<BR>
3) Use it. As an example, like so:<BR><BR>

```
class DeviceInformation extends StatelessWidget {
  DeviceInformation({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Center(
      child: ResponsiveUiWidget(
        targetOlderComputers: true,
        builder: (context, deviceInformation) {
          @override
          String toString() {
            return "Device Type: ${deviceInformation.deviceTypeInformation}\n"
                "Device OS: ${deviceInformation.deviceOS}\n"
                "Device Orientation: ${deviceInformation.orientation}\n"
                "Platform Brightness: ${deviceInformation.platformBrightness}\n"
                "Parent Widget Size: ${deviceInformation.parentWidgetSize}\n"
                "Local Widget Size (W x H): ${deviceInformation.localWidgetWidth} x ${deviceInformation.localWidgetHeight}";
          }

          return Text(
            toString(),
            style: TextStyle(fontSize: deviceInformation.deviceTypeInformation == DeviceTypeInformation.WEARABLE ? 8 : 16),
          );
        },
      ),
    );
  }
}
```

4) Another example, using ```MaterialPageRoute``` and ```PageRouteBuilder```: <BR><BR>

```
class ResponsifyApp extends StatelessWidget {
  ResponsifyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(mainAxisAlignment: MainAxisAlignment.center, children: [
        MaterialButton(
          onPressed: () {
            Navigator.push(
              context,
              MaterialPageRoute(
                builder: (context) => ResponsiveUiWidget(
                    targetOlderComputers: false,
                    builder: (context, deviceInformation) {
                      if (deviceInformation.deviceTypeInformation == DeviceTypeInformation.MOBILE) {
                        return MobileVersion();
                      } else {
                        return Scaffold(
                          body: Center(
                            child: Text("This app is compatible only with a Mobile device"),
                          ),
                        );
                      }
                    }),
              ),
            );
          },
          textColor: Colors.white,
          child: Text("Mobile"),
          color: Colors.black,
        ),
        MaterialButton(
          onPressed: () {
            Navigator.push(
              context,
              PageRouteBuilder(
                pageBuilder: (_, __, ___) => ResponsiveUiWidget(
                  targetOlderComputers: false,
                  builder: (context, deviceInformation) {
                    if (deviceInformation.deviceTypeInformation == DeviceTypeInformation.TABLET) {
                      return TabletVersion();
                    } else {
                      return Scaffold(
                        body: Center(
                          child: Text("This app is compatible only with a Tablet device"),
                        ),
                      );
                    }
                  },
                ),
              ),
            );
          },
          textColor: Colors.white,
          child: Text("Tablet"),
          color: Colors.black,
        )
      ]),
    );
  }
}

class MobileVersion extends StatelessWidget {
  const MobileVersion({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(child: Text("This is a Mobile Device")),
    );
  }
}

class TabletVersion extends StatelessWidget {
  const TabletVersion({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(child: Text("This is a Tablet Device")),
    );
  }
}
```

## Screenshots

<img src = "https://raw.githubusercontent.com/flutternoob/responsify/master/screenshots/Responsify_mobile_portrait_android.png"><BR><BR>
<img src = "https://raw.githubusercontent.com/flutternoob/responsify/master/screenshots/Responsify_mobile_landscape_android.png"><BR><BR>
<img src = "https://raw.githubusercontent.com/flutternoob/responsify/master/screenshots/Responsify_tablet_portrait_android.png"><BR><BR>
<img src = "https://raw.githubusercontent.com/flutternoob/responsify/master/screenshots/Responsify_tablet_landscape_android.png"><BR><BR>
<img src = "https://raw.githubusercontent.com/flutternoob/responsify/master/screenshots/Responsi˛fy_wearable.png"><BR><BR>
<img src = "https://raw.githubusercontent.com/flutternoob/responsify/master/screenshots/Responsify_computer_windows.png"><BR><BR>
<img src = "https://raw.githubusercontent.com/flutternoob/responsify/master/screenshots/Responsify_mobile_portrait_ios.png"><BR><BR>
<img src = "https://raw.githubusercontent.com/flutternoob/responsify/master/screenshots/Responsify_mobile_landscape_ios.png"><BR><BR>
<img src = "https://raw.githubusercontent.com/flutternoob/responsify/master/screenshots/Responsify_tablet_portrait_ios.png"><BR><BR>
<img src = "https://raw.githubusercontent.com/flutternoob/responsify/master/screenshots/Responsify_tablet_landscape_ios.png"><BR><BR>
<img src = "https://raw.githubusercontent.com/flutternoob/responsify/master/screenshots/Responsify_foldable_closed.png"><BR><BR>
<img src = "https://raw.githubusercontent.com/flutternoob/responsify/master/screenshots/Responsify_foldable_open.png"><BR>

## Note

* The benchmarks for determining thresholds for device screen sizes were determined by<BR>
  performing a Google search for minimum and maximum screen sizes for each device type<BR>
* This library has not been tested on Microsoft Surface devices, but has been tested on an emulated foldable phone<BR>
* I currently have no way to test this library on Linux devices<BR>

## Known Issues

* For a Windows computer that has dark mode enabled, the Platform brightness is identified as light mode. I do not know why.<BR>

## References

1) [https://material.io/blog/device-metrics](https://material.io/blog/device-metrics)
2) [https://developer.android.com/training/multiscreen/screendensities](https://developer.android.com/training/multiscreen/screendensities)
3) [https://www.calculatorsoup.com/calculators/technology/ppi-calculator.php](https://www.calculatorsoup.com/calculators/technology/ppi-calculator.php)
4) [https://groups.google.com/g/flutter-dev/c/oYN_prI7sio](https://groups.google.com/g/flutter-dev/c/oYN_prI7sio)
5) [https://medium.com/flutter-community/the-best-flutter-responsive-ui-pattern-ba52875d70cd](https://medium.com/flutter-community/the-best-flutter-responsive-ui-pattern-ba52875d70cd)
6) [https://stackoverflow.com/questions/49307677/how-to-get-height-of-a-widget](https://stackoverflow.com/questions/49307677/how-to-get-height-of-a-widget)
7) [https://github.com/flutter/flutter/issues/14488](https://github.com/flutter/flutter/issues/14488)

## Conclusion

I hope this library is useful for those who want an easy way to create different layouts based on device type. PRs are welcome.


CHANGELOG.md
## 1.0.0

* Initial release.

## 1.0.0+1

* Updated README.md
* Added example

## 1.0.0+2

* Fixed broken screenshots

## 1.0.0+3

* Updated README.md 

## 1.0.0+5

* Fixed grammatical errors in README.md
* Fixed library's broken Github issues link

## 1.0.1+5

* It is now possible to get the device operating system as a String
* Fixed "The '!' will have no effect because the receiver can't be null." warning
* Added iOS and Foldable phone screenshots
* Updated example and Readme with example code

## 1.0.1+6

* Updated closed foldable device screenshot

example/responsify_example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:responsify_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:responsify/responsify.dart';

void main() {
  runApp(
    MyApp(),
  );
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: "Responsive UI Demo",
      home: SafeArea(
        child: Scaffold(
          appBar: AppBar(
            backgroundColor: Colors.blue[900],
            title: Text("Responsive UI Demo"),
          ),
          body: DeviceInformation(),
        ),
      ),
    );
  }
}

///Different UI for different devices. Here the string in the Text widget and the font size
///is different for Wearables and other devices
///Different app layouts can be implemented based on device type
class DeviceInformation extends StatelessWidget {
  DeviceInformation({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Center(
      child: ResponsiveUiWidget(
        targetOlderComputers: true,
        builder: (context, deviceInformation) {
          @override
          String toString() {
            return "Device Type: ${deviceInformation.deviceTypeInformation}\n"
                "Device OS: ${deviceInformation.deviceOS}\n"
                "Device Orientation: ${deviceInformation.orientation}\n"
                "Platform Brightness: ${deviceInformation.platformBrightness}\n"
                "Parent Widget Size: ${deviceInformation.parentWidgetSize}\n"
                "Local Widget Size (W x H): ${deviceInformation.localWidgetWidth} x ${deviceInformation.localWidgetHeight}";
          }

          return Text(
            toString(),
            style: TextStyle(fontSize: deviceInformation.deviceTypeInformation == DeviceTypeInformation.WEARABLE ? 8 : 16),
          );
        },
      ),
    );
  }
}


lib/responsify_files/responsify_model.dart
import 'package:flutter/material.dart';
import 'package:responsify/responsify_files/responsify_enum.dart';

///The DeviceInformationModel is used to retrieve basic information about the device, some of which
///are available via the Flutter SDK.
class DeviceInformation {
  ///Used to determine whether the device is in Portrait or in Landscape Mode
  final Orientation? orientation;

  ///Used to determine the device type in accordance with the DeviceTypeInformation enum
  final DeviceTypeInformation? deviceTypeInformation;

  ///Used to determine the device platform
  final String? deviceOS;

  ///Used to determine the device's brightness.
  ///On Windows devices with dark mode, the Platform Brightness is incorrectly indicated as
  ///Brightness.light even with dark mode enabled on the device
  final Brightness? platformBrightness;

  ///Used to determine the size of a given widget's parent widget.
  final Size? parentWidgetSize;

  ///Used to determined the size of a widget
  final double? localWidgetWidth;
  final double? localWidgetHeight;

  ///Constructor for this class
  DeviceInformation({
    this.orientation,
    this.deviceTypeInformation,
    this.deviceOS,
    this.platformBrightness,
    this.parentWidgetSize,
    this.localWidgetWidth,
    this.localWidgetHeight,
  });
}


lib/responsify_files/responsify_ui_widget.dart
///The responsive UI Widget aids in creating responsive UIs across devices
import 'dart:io';
import 'package:flutter/material.dart';

import 'package:responsify/responsify_files/responsify_model.dart';
import 'package:responsify/responsify_files/responsify_function.dart';

///The ResponsiveUiWidget is used as a wrapper for any widget to which responsive design has to be applied
class ResponsiveUiWidget extends StatefulWidget {
  ///The builder Widget function is needed to get a widget's context and the device information
  ///based on the DeviceInformation class
  final Widget Function(BuildContext context, DeviceInformation deviceInformation)? builder;

  ///Used to target older or newer computers. If targeting older computers the value should be true
  ///If targeting newer computers, the value should be false
  final bool? targetOlderComputers;

  ///Constructor for this widget
  const ResponsiveUiWidget({Key? key, this.builder, required this.targetOlderComputers})
      : super(key: key);

  @override
  _ResponsiveUiWidgetState createState() => _ResponsiveUiWidgetState();
}

class _ResponsiveUiWidgetState extends State<ResponsiveUiWidget> {
  ///Used for local widget width and height
  double? localWidgetWidth;
  double? localWidgetHeight;

  ///Creating a GlobalKey to get the widget's current context
  final GlobalKey responsiveKey = GlobalKey();

  ///Used for determining the operating system of the device
  String devicePlatform = Platform.operatingSystem;

  @override
  void initState() {
    super.initState();

    ///Listen to the responsiveKey. Based on its context, get the width and height of the widget
    WidgetsBinding.instance.addPostFrameCallback(
        (_) => setState(() => localWidgetWidth = responsiveKey.currentContext!.size!.width));
    WidgetsBinding.instance.addPostFrameCallback(
        (_) => setState(() => localWidgetHeight = responsiveKey.currentContext!.size!.height));
  }

  @override
  Widget build(BuildContext context) {
    ///Creating a variable for getting information about the device to pass to the getDeviceType
    ///function. MediaQueryData works in a similar fashion to the media query property in CSS
    MediaQueryData mediaQueryData = MediaQuery.of(context);

    ///The LayoutBuilder widget is necessary for getting the width and height of the parent widget
    ///using the boxConstraints enum from the Size parameter
    return LayoutBuilder(
      key: responsiveKey,
      builder: (context, boxConstraints) {
        ///Passing information about the device to the DeviceInformation class constructor
        DeviceInformation deviceInformation = DeviceInformation(
          orientation: mediaQueryData.orientation,

          ///Passing the mediaQueryData parameter to the getDeviceType function to determine
          ///the device type
          deviceTypeInformation: determineDeviceType(mediaQueryData, widget.targetOlderComputers),
          deviceOS: devicePlatform,
          platformBrightness: mediaQueryData.platformBrightness,
          parentWidgetSize: Size(boxConstraints.maxWidth, boxConstraints.maxHeight),
          localWidgetWidth: localWidgetWidth,
          localWidgetHeight: localWidgetHeight,
        );

        ///Return the sizingInformation along with the widget's context
        return widget.builder!(context, deviceInformation);
      },
    );
  }
}


lib/responsify_files/responsify_function.dart
import 'dart:io';
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:responsify/responsify_files/responsify_enum.dart';

///This function is used to determine the device type, using the device's width and height in logical
///pixels, its pixel ratio, PPI and OS Platform
DeviceTypeInformation determineDeviceType(
    MediaQueryData? mediaQueryData, bool? targetOlderComputers) {
  ///Variable for device type based on the DeviceTypeInformation enum
  DeviceTypeInformation? deviceType;

  ///Device width in logical pixels
  num deviceLogicalWidth = mediaQueryData!.size.width;

  ///Device height in logical pixels
  num deviceLogicalHeight = mediaQueryData.size.height;

  ///Device pixel ratio
  num devicePixelRatio = mediaQueryData.devicePixelRatio;

  ///Diagonal length of the device in pixels calculated using Pythogoras theorem
  num? diagonalPixels;

  ///Variable used to determine device pixel density using its pixel ratio
  num? devicePixelDensity;

  ///Variable used to return the device type depending on the results of the calculation
  num? deviceSize;

  ///Function used to calculate the device size
  num calculateDeviceSize() {
    ///Online research showed that a device pixel ratio of 1 = 160 pixel density. This was used as a
    ///reference point for determining the device type. Older devices had a pixel density lower
    ///than 160 (such as 120 or 96). The following if...else block takes this into account and runs some
    ///simple calculations to return the the device size
    ///if block if device pixel ratio is > or < 1
    if (devicePixelRatio > 1 || devicePixelRatio < 1) {
      ///Device pixel resolution in width calculated by multiplying the logical pixels by the device
      ///pixel ratio
      num resolutionWidth = deviceLogicalWidth * devicePixelRatio;

      ///Device pixel resolution in height calculated by multiplying the logical pixels by the device
      ///pixel ratio
      num resolutionHeight = deviceLogicalHeight * devicePixelRatio;

      ///Calculate the diagonal pixels in accordance with Pythogoras theorem, which states that
      ///diagonal = SquareRoot((SideA^2) + (SideB^2)). Here, Side A is the width and Side B is the
      ///height. This theorem is used to determine the diagonal side of the device, as the diagonal
      ///size is the most frequent specification used by manufacturers for their devices. It can
      ///also be used in a more generic fashion to determine the device size. There is also more variety
      ///in resolution and pixel ratio across devices. Taking either of these values into account while
      ///ignoring the other is not an accurate way of determining the device size and hence the device type
      diagonalPixels = sqrt(pow(resolutionWidth, 2) + pow(resolutionHeight, 2));

      ///Calculate device pixel density = device pixel ratio * 160
      devicePixelDensity = devicePixelRatio * 160;
    }

    ///else block if device pixel ratio = 1
    else {
      ///No need to determine resolution as in the above if block when device pixel ratio = 1
      ///Device pixel ratio = 1 if device dots per inch (DPI) = 160 (baseline)
      diagonalPixels = sqrt(pow(deviceLogicalWidth, 2) + pow(deviceLogicalHeight, 2));
      devicePixelDensity = 160;
    }

    ///The formula for determining the device diagonal size in inches is the diagonal pixels/device
    ///pixel density
    return deviceSize = diagonalPixels! / devicePixelDensity!;
  }

  ///Function used to determine the device type based on the above calculations
  DeviceTypeInformation determineDeviceType() {
    ///Perform the calculations
    calculateDeviceSize();

    ///The checkHandheldDevicePlatform flag is used to determine a Handheld
    ///device's OS Platform
    bool checkHandheldDevicePlatform = Platform.isAndroid || Platform.isIOS || Platform.isFuchsia;

    ///The checkComputerPlatform flag is used to determine a Computer device's OS Platform
    ///Here a Computer may be any Desktop or Laptop/Notebook device that runs either Windows,
    ///macOS or Linux
    bool checkComputerPlatform = Platform.isWindows || Platform.isMacOS || Platform.isLinux;

    ///if...else block used to determine the device type based on the DeviceType enum and the
    ///DeviceInformation class. All threshold values are in inches
    if (deviceSize! <= 3) {
      return deviceType = DeviceTypeInformation.WEARABLE;
    } else if (deviceSize! > 3 && deviceSize! < 7.1 && checkHandheldDevicePlatform) {
      return deviceType = DeviceTypeInformation.MOBILE;
    } else if (deviceSize! > 7.1 && deviceSize! < 13.5 && checkHandheldDevicePlatform) {
      return deviceType = DeviceTypeInformation.TABLET;

      ///The baseline pixel density of 160 cannot be applied to devices whose pixel densities < 160
      ///(such as 72dpi or 96dpi).
      ///Therefore, for older computers, the above Pythogorean calculation is not taken into account
      ///as this causes a device type of UNKNOWN to be returned.
      ///Hence, DeviceTypeInformation.COMPUTER is returned based on the Platform enum and the
      ///targetOlderComputers flag for older computers.
      ///For newer computers, the Pythogorean calculation is taken into account
    } else if ((deviceSize! > 13.5 && deviceSize! < 55 && checkComputerPlatform) ||
        (targetOlderComputers! && checkComputerPlatform)) {
      return deviceType = DeviceTypeInformation.COMPUTER;
    } else {
      return deviceType = DeviceTypeInformation.UNKNOWN;
    }
  }

  ///Call the function to determine the device type
  determineDeviceType();

  ///Return the device type back to the DeviceInfoWidget
  return deviceType!;
}
